docs/gc.md

```diff
@@ -110,11 +110,45 @@ A special mode of the `gencon` policy is known as *Concurrent Scavenge* (`-Xgc:c
+
+![The diagram is explained in the surrounding text](./cr/balanced_GC_regions.png "Region structure and characteristics found in the object heap")
+
+#### Arraylets
+Most objects are easily contained within the minimum region size of 512KB. However, some large arrays might require more memory than is available in a single region. In order to support such large arrays the `balanced` collector uses an arraylet representation for large arrays. Arraylets have a spine, which contains the class pointer and size, and leaves which contain the data associated with the array. The spine also contains arrayoids which are pointers to the respective arraylet leaves as shown in the following diagram.
+
+![The diagram is explained in the surrounding text](./cr/arrayalet_diagram.png "Arraylet diagram")
+
+There are a number of advantages of using arraylets. Due to heap fragmentation over time, other collector policies might be forced to run a global garbage collection and defragmentation (compaction) phase in order to recover sufficient contiguous memory to allocate a large array. By removing the requirement that large arrays be allocated in contiguous memory, the `balanced` garbage collector is more likely to be able to satisfy such an allocation without requiring unscheduled garbage collection, and more likely still to be able to do so without a global defragmentation operation. Additionally, the `balanced` collector never needs to move an arraylet leaf once it has been allocated. The cost of relocating an array is limited to the cost of relocating the spine, so large arrays do not contribute to higher defragmentation times.
+
+Arraylets were introduced so that arrays were more cleverly stored in the heap for `balanced` and `metronome` GC policies (both are region based garbage collectors). However, there are a few drawbacks when using them. One of those is when Java Native Interface (JNI) is used. In short, JNI allows Java programs to call native code (e.g. C, C++) giving more flexibility to the program and functions that it can use. 
+
+JNI critical is used when the programmer wants direct addressability of the object. To do that, JNI needs the object to be in a contiguous region of memory, or at least make it think the object is in a contiguous region. For that to happen, JNI critical creates a temporary array on the size of total elements in the original array. It then copies element by element to this temporary array. After the JNI critical is done it copies everything back, again element by element from the temporary array to the arraylet. This operation is quite expensive. Double mapping arraylets was introduced to solve these problems.
+
+#### Double Mapping
+The initial idea of double mapping was to make a large array, which is divided across the heap in different chunks, look like a contiguous array.
```

Comment:
This makes it sound like there is a newer idea. It could be rewritten as:
The concept of double mapping is to make ...